README.md
# PcAux

---

## Archive Notice

PcAux development no longer uses this repository. So, the repository has been
archived.

Below, you can find the original README information for this repository.

---

This is the repository for the PcAux package which was formerly called "quark."

- Licensing information is given in the [LICENSE] file.
- Built tarballs of the PcAux package are available in the [builds][] directory.
- Stand-alone documentation is available in the [documentation][docs] directory.
- The source files for the most recent stable version of PcAux are available in
  the [source][src] directory.

PcAux is beta software, so please report any bugs that you encounter in the
issues section of the project page. You may also leave requests for new features
in the issues section.

Thank you for your interest in the PcAux project! I hope you find our software
useful!

## Installation
The best way to install PcAux is to use the `devtools::install_github` function.

1. First, make sure that you have **devtools** installed on your system
2. Next, execute the following lines:

        library(devtools)
        install_github("PcAux-Package/PcAux/source/PcAux")
    
3. Finally, load **PcAux** and enjoy:

        library(PcAux)

If the **devtools**-based approach does not work, you can download one of the
built tar-balls from the [builds][] directory and manually install the package
from source by executing the following lines:

        install.packages("/SAVE_PATH/PcAux_VERSION.tar.gz",
                         repos = NULL,
                         type  = "source")

Where *SAVE_PATH* is replaced by the (relative or absolute) file path to the
location where you saved the tar-ball, and *VERSION* is replaced with the correct
version number for the tar-ball that you downloaded.

## Example
A basic missing data treatment using **PcAux** might look like the following:

1. First, load and prepare your data:

        data(iris2)
        cleanData <- prepData(rawData   = iris2,
                              nomVars   = "Species",
                              ordVars   = "Petal.Width",
                              idVars    = "ID",
                              dropVars  = "Junk",
                              groupVars = "Species")

2. Next, create a set of principal component auxiliary variables:

        pcAuxOut <- createPcAux(pcAuxData = cleanData,
                                nComps    = c(3, 2))

3. Finally, use the principal component auxiliaries as the predictors in a
   multiple imputation run:

        miOut <- miWithPcAux(rawData   = iris2,
                             pcAuxData = pcAuxOut,
                             nImps     = 5)

You can also work directly with the principal component auxiliaries:

- You can merge the principal component auxiliaries back onto your raw data (e.g.,
  for use with the Graham, 2003, saturated correlates approach).

        outData <- mergePcAux(pcAuxData = pcAuxOut, rawData = iris2)

- You can also create a stand-alone predictor matrix that can be used to
  correctly incorporate the principal component auxiliaries into a separate
  MI run using the **mice** package.

        predMat <- makePredMatrix(mergedData = outData)

[builds]:  https://github.com/PcAux-Package/PcAux/tree/master/builds/
[docs]:    https://github.com/PcAux-Package/PcAux/tree/master/documentation/
[src]:     https://github.com/PcAux-Package/PcAux/tree/master/source/PcAux
[LICENSE]: https://github.com/PcAux-Package/PcAux/blob/master/LICENSE


CHANGELOG.md
# Change Log
All notable changes to the **PcAux** project will be documented in this file.

The format is based on [Keep a Changelog][kacl], and this project adheres to
[Semantic Versioning][sv].

NOTE: Changes prior to version 0.0.0.9006 were documented retrospectively and 
should, therefore, be viewed with an appropriate degree of skepticism.

## 0.0.0.9014 - 2020-04-09 - ACTIVE

### Fixed	
-Bug in check on whether `castData` is operating inside `createPcAux`
 causing `doPCA` to crash because unimputed drop variables were erroneously
 added back in
-Typo in helper function `simplePca` referencing `setNComp` instead of 	
 `setNComps`
-Bug resulting from incorrectly subsetting dropVars specified in 
 `miWithPcaux` that were not previously specified
-Bug breaking `interactType = 1` due to the absence of dummy-coded nominal 
 variables
-Bug resulting from having id variables of class `factor` with missing 
 values
-Bug in including/excluding `dropVars` between `createPcAux` and 
 `miWithPcAux`
	
## 0.0.0.9013 - 2018-09-05

### Fixed
- Bug with calculation of the cumulative proportions of variance explained by 
  the component scores 
- Bug causing inappropriate warnings when not specifying any moderators (Issue 
  #20)
- Bug causing issues when computing interaction terms with certain arrangements 
  of few moderators (Issue #18)
- Bug causing crashes when no ID variables are specified (Issue #21)
- Bug breaking `getLoggedEvents` due to inconsistent object typing by mice
	
### Changed
- Resolved final known issues with the timing utilities contributed by @ppanko
- Exported the `writeStatus` function
- Improved the way interaction terms are computed to be more robust to 
  unanticipated use-cases
- Improved implementation of the `pcAuxW` function
- Cleaned up the source code for the `PcAuxData` class definition

### Added
- A `pcAuxL` function to print licensing information
- Included @Lwaggens (Luke Waggenspack) as a contributor, particularly for help
  with locating and correcting the first bug noted in the above *fixed* 
  subsection
- A `getLoggedEvents` function used to retrieve events produced by `mice`
  during the initial imputation contributed by @ppanko 
  
## 0.0.0.9012 - 2017-11-16

### Fixed
- Changed way interactions are computed to avoid protection stack overflows 
  caused by constructing huge formulae
  
### Changed
- Fix noted in version 0.0.0.9011 partially reverted

    - Polynomials are now constructed from ordinal variables as well as 
	  continuous variables
	- It makes as much sense to construct a polynomial from an ordinal variable 
	  as it does to cast it as numeric and use it in an interaction (which we've 
	  always done)
	- Accordingly, asking for polynomials with only ordinal variables will no 
	  longer trigger an error
	  
- Added linewrapping for startup, error, and warning messages.

## 0.0.0.9011 - 2017-11-14

### Fixed
- Bug causing crashes by trying to construct polynomials with only categorical 
  variables (added an appropriate error message). 
  
## 0.0.0.9010 - 2017-10-24

### Changed
- @IMMAPbjung contributed improvements to the way that predictor matrices are 
  constructed for the initial, single imputation when *P* > *N*.
- @kylelang adjusted the way that the predictor sets for interactions and 
  polynomial terms are defined during the initial, single imputation when *P* > 
  *N*.
  
    - Polynomial terms are imputed using the same predictor set used for the 
	  untransformed version of the variable.
	- Interactions are imputed using the intersection of the predictor sets of
	  the constituent variables.
		
- Updated the maintainer email address.

## 0.0.0.9009 - 2017-09-26

### Added
- @vibhutittu contributed a modified version of *makePredMatrix* that includes
  the option to screen the PcAux variables with *mice::quickpred*

### Fixed
- Bug causing fatal errors when using parallel MI and requesting *compFormat =
  "repeated"* (Issue #14).
  
## 0.0.0.9008 - 2017-09-11

### Fixed
- Bug causing fatal errors when computing polynomial terms in *createPcAux*
- Bug causing fatal errors when merging PcAux in *miWithPcAux* with no
  non-linear component scores

## 0.0.0.9007 - 2017-06-12

### Added
- @ppanko contributed documentation for his timing/benchmarking utilities
- Included tutorial document in the "documentation" directory
- @kylelang removed *writeStatus* from the exports until the bug therewith has
  can be resolved

## 0.0.0.9006 - 2017-04-13

### Changed
- Improved the README.md file

### Added
- Added this Change Log
- Added URL and BugReports links to DESCRIPTION file

## 0.0.0.9005 - 2017-04-07

### Added
- @ppanko (Pavel Panko) contributed utilities for timing/benchmarking

## 0.0.0.9004 - 2017-03-27

### Changed
- Tightened up the dependencies by moving all but **mice** into the *imports* 
  section and using explicit namespace resolution, where feasible
 
## 0.0.0.9003 - 2017-03-26

### Fixed
- Fixed bug that was causing all PcAux scores to be used as predictors, 
  regardless of what value the user set for the *nComps* argument.

## 0.0.0.9002 - 2017-03-24

### Changed
- Adopted semantic versioning scheme suggested by [Hadley Wickham][hw]

### Fixed
- Fixed a bug that occurred while casting nominal variables that was causing 
  missing rows to be excluded from the dummy-coded representations

## 0.0.1 - 2017-03-23

### Changed
- More documentation clean-up after the package name change

## 0.0.0 - 2017-03-23

### Changed
- Changed the name of the package from **quark** to **PcAux**
- Updated README.md
- Cleaned up documentation

## <0.0.0 (i.e., Change history for **quark**) - <2017-03-23

At the same time as changing the name of this package from **quark** to 
**PcAux**, many dramatic changes were made to the package internals and to key 
aspects of the user-interface. For those early-adopters who have become 
familiar with the UI of **quark** please take some time to familiarize yourself 
with the new features of **PcAux**.


source/PcAux/R/prepData.R
### Title:        Data Preperation
### Author:       Kyle M. Lang
### Contributors: Byungkwan Jung, Pavel Panko
### Created:      2016-JAN-19
### Modified:     2018-MAY-25

##--------------------- COPYRIGHT & LICENSING INFORMATION --------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of PcAux.                                               ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General  ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program. If not, see <http://www.gnu.org/licenses/>.            ##
##----------------------------------------------------------------------------##


prepData <- function(rawData,
                     moderators = NULL,
                     nomVars    = NULL,
                     ordVars    = NULL,
                     idVars     = NULL,
                     dropVars   = NULL,
                     groupVars  = NULL,
                     simMode    = FALSE,
                     nProcess   = 1L,
                     verbose    = 2L,
                     control,
                     ...)
{
    ## Check for problems with the input values:
    if(missing(rawData)) errFun("noData")
    if(!simMode)         checkInputs()

    if(missCheck(dropVars)) dropVars <- "NONE_DEFINED"

    ## Initialize a new instance of the PcAuxData class
    ## to store all of the data and metadata for this run:
    pcAuxData <- PcAuxData(data     = rawData,
                           dropVars = dropVars,
                           simMode  = simMode,
                           nProcess = as.integer(nProcess),
                           verbose  = as.integer(verbose)
                           )

    pcAuxData$setCall(match.call(), parent = "prepData")
    pcAuxData$setTime()

    ## Make sure the control list is fully populated:
    if(!missCheck(control)) pcAuxData$setControl(x = control)

    ## Set initial machine check
    if(pcAuxData$checkStatus == "start" | pcAuxData$checkStatus == "all") pcAuxData$setStatus()

    ## Check for special variable arguments and fill the appropriate slots in the
    ## pcAuxData object:
    if(!missCheck(idVars)) {
        pcAuxData$idVars <- idVars
        pcAuxData$idCols <- as.data.frame(pcAuxData$data[ , idVars])
        pcAuxData$data   <-
            pcAuxData$data[ , setdiff(colnames(pcAuxData$data), idVars)]
    }
    if(!missCheck(groupVars))  pcAuxData$groupVars  <- groupVars
    if(!missCheck(nomVars))    pcAuxData$nomVars    <- nomVars
    if(!missCheck(ordVars))    pcAuxData$ordVars    <- ordVars
    if(!missCheck(moderators)) pcAuxData$moderators <- moderators

    pcAuxData$setTime("checkSpecial")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("checkSpecial")

### Pre-process the data ###

    ## Cast the variables to their declared types:
    castData(map = pcAuxData)

    if(!simMode) {# Don't clean data in simMode
        ## Remove constant and empty columns:
        cleanData(map = pcAuxData)

        ## Find any (bivariate) collinear variables:
        findCollin(map = pcAuxData)
    }

    pcAuxData$setTime("cast")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("cast")

    pcAuxData
}# END prepData()


source/PcAux/R/createPcAux.R
### Title:        Create Principal Component Auxiliary Variables
### Author:       Kyle M. Lang
### Contributors: Steven Chesnut, Pavel Panko, Luke Waggenspack
### Created:      2015-SEP-17
### Modified:     2020-APR-10

##--------------------- COPYRIGHT & LICENSING INFORMATION --------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of PcAux.                                               ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General  ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program. If not, see <http://www.gnu.org/licenses/>.            ##
##----------------------------------------------------------------------------##


createPcAux <- function(pcAuxData,
                        nComps,
                        interactType = 0L,
                        maxPolyPow   = 1L,
                        simMode      = FALSE,
                        seed         = NULL,
                        verbose      = 2L,
                        doImputation = TRUE,
                        castData     = !doImputation,
                        control,
                        ...)
{
    pcAuxData$setCall(match.call(), parent = "createPcAux")

    ## Set initial time and status check 
    pcAuxData$setTime()
    if(pcAuxData$checkStatus == "start" | pcAuxData$checkStatus == "all")
        pcAuxData$setStatus()

    
    ## Check for problems with the input values:
    if(missing(pcAuxData)) errFun("noPcAuxData")
    if(missing(nComps))    errFun("noNComps")
    if(!simMode)           checkInputs()
    
    ## Add elements to an extant instance of the PcAuxData class:
    pcAuxData$nComps   <- nComps
    pcAuxData$forcePmm <- FALSE # Don't give imputation options other than PMM
    pcAuxData$intMeth  <- as.integer(interactType)
    pcAuxData$maxPower <- as.integer(maxPolyPow)
    pcAuxData$simMode  <- simMode
    pcAuxData$verbose  <- as.integer(verbose)
    
    if(!missCheck(seed)) pcAuxData$seed <- as.integer(seed)
    
    ## Make sure the control list is fully populated:
    if(!missCheck(control)) pcAuxData$setControl(x = control)
    
    pcAuxData$setTime("dataCheck")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("dataCheck")
    
    ## Check for extant moderators:
    check <- missCheck(pcAuxData$moderators)
    if(check) {
        pcAuxData$moderators <- colnames(pcAuxData$data)
        if(pcAuxData$intMeth ==1 | pcAuxData$intMeth == 2)
            warnFun("noMods", map = pcAuxData)
    }

    pcAuxData$setTime("modExt")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("modExt")
    
    ## Re-cast the data if needed
    if(castData | interactType == 1) castData(map = pcAuxData)

    pcAuxData$setTime("cast")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("cast")
    
    if(doImputation) {
        ## Check for and treat any nominal variables that are missing only one
        ## datum:
        singleMissNom <-
            with(pcAuxData, (nrow(data) - respCounts == 1) &
                            (typeVec == "binary" | typeVec == "nominal")
                 )
        ## KML 2016-NOV-14: Ignore dropped variables
        singleMissNom <- setdiff(names(singleMissNom)[singleMissNom],
                                 pcAuxData$dropVars[ , 1])
        
        if(length(singleMissNom) > 0) pcAuxData$fillNomCell(singleMissNom)

        pcAuxData$setTime("doImp")
        if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("doImp")
    }
    
    ## Compute interactions for use during initial imputation:
    if(pcAuxData$intMeth == 1) {
        pcAuxData$computeInteract()
        pcAuxData$data     <- with(pcAuxData, data.frame(data, interact))
        pcAuxData$interact <- "Removed to save resources"
    }

    pcAuxData$setTime("compInt")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("compInt")
    
    ## Compute polynomials for use during initial imputation:
    if(pcAuxData$maxPower > 1) {
        pcAuxData$computePoly()
        pcAuxData$data <- with(pcAuxData, data.frame(data, poly))
        if(pcAuxData$intMeth == 1) pcAuxData$poly <- "Removed to save resources"
    }
    
    pcAuxData$setTime("compPoly")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("compPoly")
    
    ## Execute the initial, single imputation:
    if(doImputation) {

        doSingleImputation(map = pcAuxData)

        ## Use imputed data to update nominal variable representations:
        if(length(pcAuxData$nomVars) > 0) pcAuxData$codeNomVars()
   
        pcAuxData$setTime("doSingle")
        if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("doSingle")
    }
    
    ## Extract the linear principal component scores:
    doPCA(map = pcAuxData)

    pcAuxData$setTime("doPCA")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("doPCA")
    
    ## Are we constructing seperate non-linear PcAux?
    if(pcAuxData$nComps[2] != 0) {
        ## Undo dummy coding to facilitate interaction calculation:
        if(!missCheck(pcAuxData$nomVars))
            pcAuxData$castNomVars(toNumeric = FALSE)
        
        ## Construct and orthogonalize interaction terms:
        if(pcAuxData$intMeth > 1) pcAuxData$computeInteract()
        
        ## Extract the nonlinear principal component scores:
        doPCA(map = pcAuxData)

        pcAuxData$setTime("doNLinear")
        if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("doNLinear")
    }

    ## Remove unnecessary representation of nominal variables:
    pcAuxData$facNoms <- "Removed to save resources"
    pcAuxData$dumNoms <- "Removed to save resources"

    pcAuxData$setTime("rmVars")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("rmVars")
    
    pcAuxData
}# END createPcAux()


source/PcAux/R/miWithPcAux.R
### Title:        Conduct Multiple Imputation with PC Auxiliaries
### Author:       Kyle M. Lang
### Contributors: Steven Chesnut, Pavel Panko, Luke Waggenspack
### Created:      2015-SEP-17
### Modified:     2020-APR-09
### Purpose:      Use the principal component auxiliaries produced by
###               createPcAux() to conduct MI.

##--------------------- COPYRIGHT & LICENSING INFORMATION --------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of PcAux.                                               ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General  ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program. If not, see <http://www.gnu.org/licenses/>.            ##
##----------------------------------------------------------------------------##


miWithPcAux <- function(rawData,
                        pcAuxData,
                        nImps      = 100L,
                        nomVars    = NULL,
                        ordVars    = NULL,
                        idVars     = NULL,
                        dropVars   = "useExtant",
                        nComps     = NULL,
                        compFormat = "list",
                        seed       = NULL,
                        simMode    = FALSE,
                        forcePmm   = FALSE,
                        nProcess   = 1L,
                        verbose    = 2L,
                        control)
{
    pcAuxData$setCall(match.call(), parent = "miWithPcAux")

    pcAuxData$setTime()
    if(pcAuxData$checkStatus == "start" | pcAuxData$checkStatus == "all") pcAuxData$setStatus()
    
    if(missing(rawData))   errFun("noData")
    if(missing(pcAuxData)) errFun("noPcAuxData")
    
    ## Get variable types:
    if(!missCheck(nomVars)) {
        pcAuxData$nomVars  <- nomVars
        removeNoms         <- which(pcAuxData$dropVars[,1] %in% nomVars)
        pcAuxData$dropVars <- pcAuxData$dropVars[-removeNoms, ]
    }
    if(!missCheck(ordVars)) {
        pcAuxData$ordVars  <- ordVars
        removeOrds         <- which(pcAuxData$dropVars[,1] %in% ordVars)
        pcAuxData$dropVars <- pcAuxData$dropVars[-removeOrds, ]
    }
    if(!missCheck(idVars)) {
        pcAuxData$idVars   <- idVars
        removeIds         <- which(pcAuxData$dropVars[,1] %in% idVars)
        pcAuxData$dropVars <- pcAuxData$dropVars[-removeIds, ]
    }
    if(length(dropVars) == 1 && dropVars == "useExtant") {
        tmp <- pcAuxData$dropVars[pcAuxData$dropVars[ , 2] == "user_defined", ]
        if(class(tmp) != "matrix") tmp <- matrix(tmp, 1, 2)
        pcAuxData$dropVars <- tmp
    } else if(!missCheck(dropVars)) {
        pcAuxData$dropVars <- cbind(dropVars, "user_defined")
        pcAuxData$nomVars  <- setdiff(pcAuxData$nomVars, dropVars)
        pcAuxData$ordVars  <- setdiff(pcAuxData$ordVars, dropVars)
        pcAuxData$idVars   <- setdiff(pcAuxData$idVars, dropVars)
    } else {
        pcAuxData$dropVars <- cbind("NONE_DEFINED", "user_defined")
    }

    ## Check inputs' validity:
    if(!simMode) checkInputs()

    pcAuxData$setTime("varTypes")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("varTypes")
    
    ## Combine the principal component auxiliaries with the raw data:
    mergePcAux(pcAuxData = pcAuxData,
               rawData   = rawData,
               nComps    = nComps,
               intern    = TRUE)

    pcAuxData$setTime("mergePcAux")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("mergePcAux")
    
    ## Populate new fields in the extant PcAuxData object:
    pcAuxData$nImps      <- as.integer(nImps)
    pcAuxData$simMode    <- simMode
    pcAuxData$compFormat <- compFormat
    pcAuxData$forcePmm   <- forcePmm
    pcAuxData$verbose    <- as.integer(verbose)
    
    if(!missCheck(seed)) pcAuxData$seed <- as.integer(seed)
    
    ## Make sure the control list is fully populated:
    if(!missCheck(control)) {
        if(!is.null(control$minPredCor))
            control$minPredCor <- c(pcAuxData$minPredCor[1], control$minPredCor)
        pcAuxData$setControl(x = control)
    }
    
    pcAuxData$setTime("popNew")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("popNew")
    
    ## Cast the variables to their appropriate types:
    castData(map = pcAuxData)
    
    ## Check and clean the data:
    if(!simMode) cleanData(map = pcAuxData)

    pcAuxData$setTime("reCast")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("reCast")
    
    ## Check for and treat any single nominal variables that are missing
    ## only one datum
    singleMissNom <- with(pcAuxData, (nrow(data) - respCounts == 1) &
                                     (typeVec == "binary" | typeVec == "nominal")
                          )
    if(any(singleMissNom))
        pcAuxData$fillNomCell(colnames(pcAuxData$data[singleMissNom]))

    pcAuxData$setTime("nomMis")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("nomMis")
    
    if(verbose > 1) cat("\nMultiply imputing missing data...\n")
    
    ## Construct a predictor matrix for mice():
    if(verbose > 1) cat("--Constructing predictor matrix...")
    nLin    <- length(grep("^linPC\\d", colnames(pcAuxData$data)))
    nNonLin <- ifelse(nComps[2] == 0,
                      0,
                      length(grep("^nonLinPC\\d", colnames(pcAuxData$data)))
                      )
    
    ## Make sure the quickpred procedure is well-defined (if requested):
    check <- pcAuxData$useQuickPred && length(pcAuxData$minPredCor) == 1 
    if(check) errFun("noMinCor")
    
    predMat <- makePredMatrix(mergedData   = pcAuxData$data,
                              nLinear      = nLin,
                              nNonLinear   = nNonLin,
                              useQuickPred = pcAuxData$useQuickPred,
                              minCor       = ifelse(pcAuxData$useQuickPred,
                                                    pcAuxData$minPredCor[2],
                                                    NULL)
                              )
    if(verbose > 1) cat("done.\n")

    pcAuxData$setTime("predMat")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("predMat")
    
    ## Specify a vector of elementary imputation methods:
    if(verbose > 1) cat("--Creating method vector...")
    pcAuxData$createMethVec()
    if(verbose > 1) cat("done.\n")

    pcAuxData$setTime("methVec")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("methVec")
    
    if(forcePmm & verbose > 1) cat("PMM forced by user.\n")
    
    ## Multiply impute the missing data in 'mergedData':
    if(verbose > 1) cat("--Imputing missing values...\n")
    
    if(nProcess == 1) {# Impute in serial
        pcAuxData$miceObject <- try(
            mice(pcAuxData$data,
                 m               = nImps,
                 maxit           = 1L,
                 predictorMatrix = predMat,
                 method          = pcAuxData$methVec,
                 printFlag       = verbose == 2,
                 ridge           = pcAuxData$miceRidge,
                 seed            = pcAuxData$seed,
                 nnet.MaxNWts    = pcAuxData$maxNetWts),
            silent = TRUE)

        pcAuxData$setTime("impSerial")
        if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("impSerial")
        
        if(class(pcAuxData$miceObject) != "try-error") {
            pcAuxData$data <- "Removed to save resources."
            
            ## Complete the incomplete data sets:
            pcAuxData$completeMiData()

            pcAuxData$setTime("completeMi")
            if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("completeMi")
            
        } else {
            errFun("miceCrash", map = pcAuxData)
        }
        
    } else {# Impute in parallel
        myCluster <- makeCluster(nProcess)
        clusterEvalQ(myCluster, library(mice))
        
        pcAuxData$miDatasets <- parLapply(myCluster,
                                          X       = c(1 : nImps),
                                          fun     = parallelMice,
                                          predMat = predMat,
                                          map     = pcAuxData)
        
        stopCluster(myCluster)

        pcAuxData$setTime("impParallel")
        if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("impParallel")
        
        pcAuxData$transformMiData()

        pcAuxData$setTime("transformMI")
        if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("transformMI")
        
    }
    
    if(verbose > 1) cat("\n--done.\n")
    if(verbose > 1) cat("Complete.\n")

    pcAuxData$setTime("miEnd")
    if(pcAuxData$checkStatus == "all") pcAuxData$setStatus("miEnd")
    
    pcAuxData
}# END miWithPcAux()
    
